db/migrate/20200226173054_populate_new_option_node_id_columns.rb

```diff
@@ -0,0 +1,134 @@
+    if (pairs = missing_pairs).any?
+      if File.exist?(FLAG_PATH)
+        create_missing_nodes(pairs)
+        copy_option_id_to_option_node_id
+      else
+        print_missing_pairs(pairs)
+        raise Exception
+      end
+    end
+    assert_no_missing_pairs_or_non_matching_option_ids
+  end
+
+  private
+
+  def copy_option_id_to_option_node_id
+    # This trigger is slow and the search data doesn't need to be updated here.
+    execute("ALTER TABLE answers DISABLE TRIGGER answers_before_insert_update_row_tr")
+    puts "Copying for answers..."
+    execute_partitioned(answer_update_query, "answers")
+    puts "Copying for choices..."
+    execute_partitioned(choice_update_query, "choices")
+    execute("ALTER TABLE answers ENABLE TRIGGER answers_before_insert_update_row_tr")
+  end
+
+  def answer_update_query
+    <<-SQL
+      UPDATE answers SET option_node_id = option_nodes.id
+        FROM option_nodes, questions, form_items
+        WHERE __PARTITION__
+          AND answers.option_id IS NOT NULL
+          AND answers.option_node_id IS NULL
+          AND answers.questioning_id = form_items.id
+          AND form_items.question_id = questions.id
+          AND questions.option_set_id = option_nodes.option_set_id
+          AND option_nodes.option_id = answers.option_id
+    SQL
+  end
+
+  def choice_update_query
+    <<-SQL
+      UPDATE choices SET option_node_id = option_nodes.id
+        FROM answers, option_nodes, questions, form_items
+        WHERE __PARTITION__
+          AND choices.option_id IS NOT NULL
+          AND choices.option_node_id IS NULL
+          AND choices.answer_id = answers.id
+          AND answers.questioning_id = form_items.id
+          AND form_items.question_id = questions.id
+          AND questions.option_set_id = option_nodes.option_set_id
+          AND option_nodes.option_id = choices.option_id
+    SQL
+  end
+
+  def execute_partitioned(query, table, parts = 100)
+    per_part = 16**32 / parts
+    max = 16**32 - 1
+    (0...parts).to_a.each do |num|
+      puts "Part #{num + 1}/#{parts}"
+      lower = (num * per_part).to_s(16).rjust(32, "0")
+      upper = (num == parts - 1 ? max : (num + 1) * per_part).to_s(16).rjust(32, "0")
+      clause = "#{table}.id >= '#{lower}' AND #{table}.id < '#{upper}'"
+      execute(query.sub("__PARTITION__", clause))
+    end
+  end
+
+  def missing_pairs
+    null_answers = Answer.where(option_node_id: nil).where.not(option_id: nil)
+      .includes(form_item: :question)
+    null_choices = Choice.where(option_node_id: nil).where.not(option_id: nil)
+      .includes(answer: {form_item: :question})
+    return [] if null_answers.none? && null_choices.none?
+
+    pairs = null_answers.map { |a| [a.option_set_id, a.option_id] }
+    pairs.concat(null_choices.map { |c| [c.answer.option_set_id, c.option_id] })
+  end
+
+  def print_missing_pairs(pairs)
+    puts "Some answers/choices refer to options not contained in the expected option set:\n"
```

Comment:
I ran into this while migrating locally; it's unclear at a glance when this is a problem or not. Can you clarify how to know if a manual fix is needed vs. when to just touch the flag?

Is this related to the security checks we recently added to prevent submitting invalid options?

Reply:
Kind of related to the security fix.

Manual fix is needed if it seems like high stakes data. I'll add more text to clarify:

```
    puts "If the option set is not in active use, feel free to use the automated fix."
    puts "If the option set is in active use, consider the effects of adding new options to it."
    puts "You may prefer to manually delete the answers/choices in question.
```